oldschnorr.go

```diff
@@ -0,0 +1,148 @@
+package secp256k1
+
+// #include "./depend/secp256k1/include/secp256k1_oldschnorr.h"
+import "C"
+import "github.com/pkg/errors"
+
+// SchnorrPublicKey is a PublicKey type used to sign and verify Schnorr signatures.
+// The struct itself is an opaque data type that should only be created via the supplied methods.
+type SchnorrPublicKey struct {
+	pubkey C.secp256k1_pubkey
+}
+
+// SchnorrSignature is a type representing a Schnorr Signature.
+// The struct itself is an opaque data type that should only be created via the supplied methods.
+type SchnorrSignature struct {
+	signature [64]byte
+}
+
+// Serialize returns a 64 byte serialized signature
+func (signature *SchnorrSignature) Serialize() [64]byte {
+	return signature.signature
+}
+
+// ParseSchnorrSignature parses a 64 byte serialized schnorr signature into a SchnorrSignature type.
+func ParseSchnorrSignature(serializedSignature [64]byte) SchnorrSignature {
+	signature := SchnorrSignature{}
+	signature.signature = serializedSignature
+	return signature
+}
+
+// SchnorrPublicKey generates a PublicKey for the corresponding private key.
+func (key *PrivateKey) SchnorrPublicKey() (*SchnorrPublicKey, error) {
+	pubkey := SchnorrPublicKey{}
+	cPtrPrivateKey := (*C.uchar)(&key.privateKey[0])
+	ret := C.secp256k1_ec_pubkey_create(context, &pubkey.pubkey, cPtrPrivateKey)
+	if ret != 1 {
+		return nil, errors.New("failed Generating an SchnorrPublicKey. You should call `ParsePrivateKey` before calling this")
+	}
+	return &pubkey, nil
+}
+
+// SchnorrSign creates a schnorr signature using the private key and the input hashed message.
+// Notice: the [32] byte array *MUST* be a hash of a message.
+func (key *PrivateKey) SchnorrSign(hash [32]byte) (*SchnorrSignature, error) {
+	signature := SchnorrSignature{}
+	cPtrSig := (*C.uchar)(&signature.signature[0])
+	cPtrHash := (*C.uchar)(&hash[0])
+	cPtrPrivKey := (*C.uchar)(&key.privateKey[0])
+	ret := C.secp256k1_schnorr_sign(context, cPtrSig, cPtrHash, cPtrPrivKey, nil, nil)
+	if ret != 1 {
+		return nil, errors.New("failed Signing. You should call `ParsePrivateKey` before calling this")
+	}
+	return &signature, nil
+}
+
+// SchnorrVerify verifies a schnorr signature using the public key and the input hashed message.
+// Notice: the [32] byte array *MUST* be a hash of a message you hashed yourself.
+func (key *SchnorrPublicKey) SchnorrVerify(hash [32]byte, signature SchnorrSignature) bool {
```

Comment:
Please add type `Hash [32]byte` and use pointer to it instead of direct `[32]byte`.
Also use it wherever ecmh returns a hash, and define similar type for ecmh's [64]byte serialization format.

Reply:
What does that help/give?

Reply:
Than you will be able to pass by reference easily.

Reply:
Done